File: src/test/java/com/github/njuro/jard/ban/BanIntegrationTest.java
package com.github.njuro.jard.ban;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import com.github.njuro.jard.common.Mappings;
import com.github.njuro.jard.common.MockRequestTest;
import com.github.njuro.jard.common.WithMockUserAuthorities;
import com.github.njuro.jard.user.UserAuthority;
import com.github.njuro.jard.user.UserFacade;
import com.github.njuro.jard.user.UserForm;
import com.github.njuro.jard.user.UserRole;
import java.time.OffsetDateTime;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.SpyBean;
import org.springframework.http.HttpMethod;
import org.springframework.transaction.annotation.Transactional;

@SpringBootTest
@Transactional
class BanIntegrationTest extends MockRequestTest {

  private static final String API_ROOT = Mappings.API_ROOT_BANS;

  @Autowired private BanFacade banFacade;
  @SpyBean private UserFacade userFacade;

  private BanForm banForm;

  @BeforeEach
  void setUp() {
    banForm =
        BanForm.builder()
            .ip("127.0.0.1")
            .reason("Test")
            .validTo(OffsetDateTime.now().plusDays(1))
            .build();

    var user =
        userFacade.createUser(
            UserForm.builder()
                .username("user")
                .password("password")
                .passwordRepeated("password")
                .email("user@email.com")
                .role(UserRole.ADMIN)
                .build());
    when(userFacade.getCurrentUser()).thenReturn(user);
  }

  @Test
  @WithMockUserAuthorities(value = {UserAuthority.MANAGE_BANS, UserAuthority.VIEW_IP})
  void testCreateBan() throws Exception {
    assertThat(banFacade.getActiveBan(banForm.getIp())).isNull();

    performMockRequest(HttpMethod.PUT, API_ROOT, banForm)
        .andExpect(status().isOk())
        .andExpect(nonEmptyBody());

    assertThat(banFacade.getActiveBan(banForm.getIp())).isNotNull();
  }

  @Test
  void testGetOwnBan() throws Exception {
    banFacade.createBan(banForm);

    var result =
        performMockRequest(HttpMethod.GET, API_ROOT + "/me")
            .andExpect(status().isOk())
            .andExpect(nonEmptyBody())
            .andReturn();

    assertThat(getResponse(result, Ban.class))
        .extracting(Ban::getReason)
        .isEqualTo(banForm.getReason());
  }

  @Test
  @WithMockUserAuthorities(value = {UserAuthority.MANAGE_BANS, UserAuthority.VIEW_IP})
  void testGetAllBans() throws Exception {
    String originalIp = banForm.getIp();
    banFacade.createBan(banForm);
    String newIp = "127.0.1.1";
    banForm.setIp(newIp);
    banFacade.createBan(banForm);

    var result =
        performMockRequest(HttpMethod.GET, API_ROOT)
            .andExpect(status().isOk())
            .andExpect(nonEmptyBody())
            .andReturn();

    assertThat(getResponseCollection(result, List.class, Ban.class))
        .extracting(Ban::getIp)
        .containsExactlyInAnyOrder(originalIp, newIp);
  }

  @Test
  @WithMockUserAuthorities(value = {UserAuthority.MANAGE_BANS, UserAuthority.VIEW_IP})
  void testEditBan() throws Exception {
    var ban = banFacade.createBan(banForm);

    assertThat(banFacade.resolveBan(ban.getId()))
        .isNotNull()
        .extracting(Ban::getReason)
        .isEqualTo(banForm.getReason());

    String newReason = "Spam";
    banForm.setReason(newReason);
    performMockRequest(
            HttpMethod.POST,
            buildUri(API_ROOT + Mappings.PATH_VARIABLE_BAN + "/edit", ban.getId()),
            banForm)
        .andExpect(status().isOk())
        .andExpect(nonEmptyBody());

    assertThat(banFacade.resolveBan(ban.getId()))
        .isNotNull()
        .extracting(Ban::getReason)
        .isEqualTo(newReason);
  }

  @Test
  @WithMockUserAuthorities(value = {UserAuthority.MANAGE_BANS, UserAuthority.VIEW_IP})
  void testUnban() throws Exception {
    var ban = banFacade.createBan(banForm);

    assertThat(banFacade.resolveBan(ban.getId()))
        .isNotNull()
        .extracting(Ban::getStatus)
        .isEqualTo(BanStatus.ACTIVE);

    var unbanForm = UnbanForm.builder().ip(ban.getIp()).reason("OK").build();
    performMockRequest(
            HttpMethod.POST,
            buildUri(API_ROOT + Mappings.PATH_VARIABLE_BAN + "/unban", ban.getId()),
            unbanForm)
        .andExpect(status().isOk())
        .andExpect(nonEmptyBody());

    assertThat(banFacade.resolveBan(ban.getId()))
        .isNotNull()
        .extracting(Ban::getStatus)
        .isEqualTo(BanStatus.UNBANNED);
  }
}

File: src/test/kotlin/com/github/njuro/jard/ban/BanIntegrationTest.kt
package com.github.njuro.jard.ban

import com.github.njuro.jard.MockMvcTest
import com.github.njuro.jard.TestDataRepository
import com.github.njuro.jard.WithContainerDatabase
import com.github.njuro.jard.WithMockJardUser
import com.github.njuro.jard.ban
import com.github.njuro.jard.ban.dto.BanDto
import com.github.njuro.jard.ban.dto.BanForm
import com.github.njuro.jard.common.Mappings
import com.github.njuro.jard.toForm
import com.github.njuro.jard.toUnbanForm
import com.github.njuro.jard.user.UserAuthority
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.optional.shouldBePresent
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post
import org.springframework.test.web.servlet.put
import org.springframework.transaction.annotation.Transactional
import java.time.OffsetDateTime
import java.util.UUID

@WithContainerDatabase
@Transactional
internal class BanIntegrationTest : MockMvcTest() {

    @Autowired
    private lateinit var db: TestDataRepository

    @Nested
    @DisplayName("create ban")
    @WithMockJardUser(UserAuthority.MANAGE_BANS, UserAuthority.VIEW_IP)
    inner class CreateBan {
        private fun createBan(banForm: BanForm) = mockMvc.post(Mappings.API_ROOT_BANS) { body(banForm) }

        @Test
        fun `create valid ban`() {
            createBan(ban().toForm()).andExpect { status { isCreated() } }.andReturnConverted<BanDto>()
                .shouldNotBeNull()
        }

        @Test
        fun `don't create invalid ban`() {
            createBan(ban(ip = "123.456").toForm()).andExpect { status { isBadRequest() } }
        }
    }

    @Nested
    @DisplayName("get own ban")
    inner class GetOwnBan {
        private fun getOwnBan(ip: String) =
            mockMvc.get("${Mappings.API_ROOT_BANS}/me") { setUp(); with { it.apply { remoteAddr = ip } } }

        @Test
        fun `active ban exists`() {
            val ban = db.insert(ban(ip = "127.0.0.1"))
            getOwnBan(ban.ip).andExpect { status { isOk() } }.andReturnConverted<BanDto>().shouldNotBeNull()
        }

        @Test
        fun `active ban doesn't exists`() {
            getOwnBan("127.0.0.1").andExpect { status { isOk() } }.andReturn().response.contentLength shouldBe 0
        }
    }

    @Test
    @WithMockJardUser(UserAuthority.MANAGE_BANS, UserAuthority.VIEW_IP)
    fun `get all bans`() {
        (1..3).forEach { db.insert(ban(ip = "127.0.0.$it")) }
        mockMvc.get(Mappings.API_ROOT_BANS) { setUp() }.andExpect { status { isOk() } }
            .andReturnConverted<List<BanDto>>() shouldHaveSize 3
    }

    @Nested
    @DisplayName("edit ban")
    @WithMockJardUser(UserAuthority.MANAGE_BANS, UserAuthority.VIEW_IP)
    inner class EditBan {
        private fun editBan(id: UUID, editForm: BanForm) =
            mockMvc.put("${Mappings.API_ROOT_BANS}/$id") { body(editForm) }

        @Test
        fun `edit valid ban`() {
            val ban = db.insert(ban())
            val editForm = ban.toForm().apply { reason = "Updated reason" }

            editBan(ban.id, editForm).andExpect { status { isOk() } }.andReturnConverted<BanDto>().shouldNotBeNull()
            db.select(ban).shouldBePresent { it.reason shouldBe editForm.reason }
        }

        @Test
        fun `don't edit non-existing ban`() {
            editBan(UUID.randomUUID(), ban().toForm()).andExpect { status { isNotFound() } }
        }

        @Test
        fun `don't edit invalid ban`() {
            val ban = db.insert(ban())
            val editForm = ban.toForm().apply { validTo = OffsetDateTime.now().minusDays(1) }

            editBan(ban.id, editForm).andExpect { status { isBadRequest() } }
        }
    }

    @Nested
    @DisplayName("unban")
    @WithMockJardUser(UserAuthority.MANAGE_BANS, UserAuthority.VIEW_IP)
    inner class Unban {
        private fun unban(id: UUID, unbanForm: UnbanForm) =
            mockMvc.put("${Mappings.API_ROOT_BANS}/$id/unban") { body(unbanForm) }

        @Test
        fun `valid unban`() {
            val ban = db.insert(ban(status = BanStatus.ACTIVE))

            unban(ban.id, ban.toUnbanForm(unbanReason = "OK")).andExpect { status { isOk() } }
                .andReturnConverted<BanDto>().shouldNotBeNull()
            db.select(ban).shouldBePresent { it.status shouldBe BanStatus.UNBANNED }
        }

        @Test
        fun `non-existing ban`() {
            unban(UUID.randomUUID(), ban().toUnbanForm()).andExpect { status { isNotFound() } }
        }

        @Test
        fun `invalid unban`() {
            val ban = db.insert(ban(status = BanStatus.ACTIVE))

            unban(ban.id, ban.toUnbanForm().apply { ip = "123.456" }).andExpect { status { isBadRequest() } }
            db.select(ban).shouldBePresent { it.status shouldBe BanStatus.ACTIVE }
        }
    }
}
